contracts/src/locker/token_locker.cairo

```diff
@@ -0,0 +1,336 @@
+    use core::traits::TryInto;
+    use debug::PrintTrait;
+    use openzeppelin::token::erc20::{ERC20ABIDispatcher, ERC20ABIDispatcherTrait};
+    use starknet::{
+        ContractAddress, contract_address_const, get_caller_address, get_contract_address,
+        get_block_timestamp, Store
+    };
+    use unruggable::locker::errors;
+
+
+    #[storage]
+    struct Storage {
+        min_lock_time: u64,
+        lock_nonce: u128,
+        locks: LegacyMap<u128, TokenLock>,
+        user_locks: LegacyMap<ContractAddress, List<u128>>,
+    }
+
+
+    /// Events
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        TokenLocked: TokenLocked,
+        TokenUnlocked: TokenUnlocked,
+        TokenWithdrawn: TokenWithdrawn,
+        LockOwnershipTransferred: LockOwnershipTransferred,
+        LockDurationIncreased: LockDurationIncreased,
+        LockAmountIncreased: LockAmountIncreased
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct TokenLocked {
+        #[key]
+        lock_id: u128,
+        token: ContractAddress,
+        owner: ContractAddress,
+        amount: u256,
+        unlock_time: u64,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct TokenUnlocked {
+        #[key]
+        lock_id: u128,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct TokenWithdrawn {
+        #[key]
+        lock_id: u128,
+        amount: u256
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct LockOwnershipTransferred {
+        #[key]
+        lock_id: u128,
+        new_owner: ContractAddress,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct LockDurationIncreased {
+        #[key]
+        lock_id: u128,
+        new_unlock_time: u64,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct LockAmountIncreased {
+        lock_id: u128,
+        amount_to_increase: u256
+    }
+
+    #[derive(Drop, Copy, starknet::Store, Serde, PartialEq)]
+    struct TokenLock {
+        token: ContractAddress,
+        owner: ContractAddress,
+        amount: u256,
+        unlock_time: u64,
+    }
+
+    /// Initializes a new instance of a TokenLocker contract.  The constructor
+    /// sets the minimum lock time for all locks created by this contract.
+    ///
+    /// # Arguments
+    ///
+    /// * `min_lock_time` - The minimum lock time applicable to all locks
+    ///  created by this contract.
+    #[constructor]
+    fn constructor(ref self: ContractState, min_lock_time: u64,) {
+        self.min_lock_time.write(min_lock_time);
+    }
+
+    #[abi(embed_v0)]
+    impl TokenLocker of unruggable::locker::ITokenLocker<ContractState> {
+        fn lock_tokens(
+            ref self: ContractState,
+            token: ContractAddress,
+            amount: u256,
+            unlock_time: u64,
+            withdrawer: ContractAddress
+        ) -> u128 {
+            //TODO(safety): Add a check to verify that the token locked is an LP pair - to avoid people locking tokens
```

Comment:
We could call the is_memecoin function from the factory here. It would be sufficient imo

Reply:
something like factory.is_memecoin(pair.token(0))

Reply:
otherwise LGTM

Reply:
imo out of scope for MVP but its just a good idea for later - hence I keep it as todo so that we can get back on it later if we want